homeassistant/components/homeconnect/api.py

```diff
@@ -0,0 +1,390 @@
+
+from asyncio import run_coroutine_threadsafe
+import logging
+
+import homeconnect
+
+from homeassistant import config_entries, core
+from homeassistant.core import callback
+from homeassistant.helpers import config_entry_oauth2_flow
+from homeassistant.helpers.entity import Entity
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ConfigEntryAuth(homeconnect.HomeConnectAPI):
+    """Provide Home Connect authentication tied to an OAuth2 based config entry."""
+
+    def __init__(
+        self,
+        hass: core.HomeAssistant,
+        config_entry: config_entries.ConfigEntry,
+        implementation: config_entry_oauth2_flow.AbstractOAuth2Implementation,
+    ):
+        """Initialize Home Connect Auth."""
+        self.hass = hass
+        self.config_entry = config_entry
+        self.session = config_entry_oauth2_flow.OAuth2Session(
+            hass, config_entry, implementation
+        )
+        super().__init__(self.session.token)
+
+    def refresh_tokens(self) -> dict:
+        """Refresh and return new Home Connect tokens using Home Assistant OAuth2 session."""
+        run_coroutine_threadsafe(
+            self.session.async_ensure_token_valid(), self.hass.loop
+        ).result()
+
+        return self.session.token
+
+    def get_devices(self):
+        """Get a dictionary of devices."""
+        appl = self.get_appliances()
+        devices = []
+        for app in appl:
+            if app.type == "Dryer":
+                device = Dryer(app)
+            elif app.type == "Washer":
+                device = Washer(app)
+            elif app.type == "Dishwasher":
+                device = Dishwasher(app)
+            elif app.type == "FridgeFreezer":
+                device = FridgeFreezer(app)
+            elif app.type == "Oven":
+                device = Oven(app)
+            elif app.type == "CoffeeMaker":
+                device = CoffeeMaker(app)
+            elif app.type == "Hood":
+                device = Hood(app)
+            elif app.type == "Hob":
+                device = Hob(app)
+            else:
+                _LOGGER.warning("Appliance type %s not implemented.", app.type)
+                continue
+            devices.append({"device": device, "entities": device.get_entities()})
+        return devices
+
+
+class HomeConnectDevice:
+    """Generic Home Connect device."""
+
+    # for some devices, this is instead 'BSH.Common.EnumType.PowerState.Standby'
+    # see https://developer.home-connect.com/docs/settings/power_state
+    power_off_state = "BSH.Common.EnumType.PowerState.Off"
+
+    def __init__(self, appliance):
+        """Initialize the device."""
+        from homeconnect.api import HomeConnectError
+
+        self.appliance = appliance
+        try:
+            self.appliance.get_status()
+        except (HomeConnectError, ValueError):
+            _LOGGER.debug("Unable to fetch appliance status. Probably offline.")
```

Comment:
Shouldn't this be a `warning` instead of a `debug` log?
This way a user is not automatically informed when something goes wrong e.g. an exception occurs.

Reply:
This occurs when the appliance is offline during start up, e.g. when the washing machine is switched off, which is a pretty generic/likely scenario, so I wouldn't warn the user about it. Once it comes online (is switched on), the SSE stream should report the necessary updates.

Reply:
Okay that sounds very reasonable to me, was just curious why you used debug in these cases.